Sources_v3/StreamChat/APIClient/APIClient.swift

```diff
@@ -63,4 +65,52 @@ class APIClient {
             task.resume()
         }
     }
+
+    func uploadFile<Response: Decodable>(
+        endpoint: Endpoint<Response>,
+        multipartFormData: MultipartFormData,
+        progress: ((Double) -> Void)? = nil,
+        completion: @escaping (Result<Response, Error>) -> Void
+    ) {
+        encoder.encodeRequest(for: endpoint) { [unowned self] (requestResult) in
+            var urlRequest: URLRequest
+            do {
+                urlRequest = try requestResult.get()
+            } catch {
+                log.error(error)
+                completion(.failure(error))
+                return
+            }
+
+            let data = multipartFormData.multipartFormData
+            urlRequest.addValue("multipart/form-data; boundary=\(multipartFormData.boundary)", forHTTPHeaderField: "Content-Type")
+            urlRequest.httpBody = data
+
+            let task = self.session.dataTask(with: urlRequest) { [decoder = self.decoder] (data, response, error) in
+                do {
+                    let decodedResponse: Response = try decoder.decodeRequestResponse(
+                        data: data,
+                        response: response,
+                        error: error
+                    )
+                    completion(.success(decodedResponse))
+                } catch {
+                    completion(.failure(error))
+                }
+            }
+
+            if let progressListener = progress {
+                let taskID = task.taskIdentifier
+                self.taskProgressObservers[taskID] = task.progress.observe(\.fractionCompleted) { [weak self] progress, _ in
```

Comment:
`self.taskProgressObservers[taskID] =` is not an atomic operation. You need to use `self.self.taskProgressObservers { obseervers in ... }`

Reply:
fixed in 5d1d4587fc27b39cd66541f4ac290cfee037b9a6